gadget-container/gadgets/collector-process/c/ebpf.c

```diff
@@ -0,0 +1,106 @@
+#include <linux/bpf.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+
+struct bpf_iter_meta {
+	__bpf_md_ptr(struct seq_file *, seq);
+	__u64 session_id;
+	__u64 seq_num;
+};
+
+struct bpf_iter__task {
+	struct bpf_iter_meta *meta;
+	struct task_struct *task;
+} __attribute__((preserve_access_index));
+
+struct ns_common {
+	unsigned int inum;
+} __attribute__((preserve_access_index));
+
+struct nsproxy {
+	struct mnt_namespace *mnt_ns;
+} __attribute__((preserve_access_index));
+
+struct task_struct {
+	int pid;
+	int tgid;
+	char comm[16];
+	struct nsproxy *nsproxy;
+} __attribute__((preserve_access_index));
+
+struct mnt_namespace {
+	struct ns_common ns;
+}__attribute__((preserve_access_index));
+
+struct container {
+	char ContainerID[256];
+	char KubernetesNamespace[256];
+	char KubernetesPod[256];
+	char KubernetesContainer[256];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, __u64);
+	__type(value, struct container);
+	__uint(max_entries, 1024);
+} containers SEC(".maps");
+
+
+/* From: tools/lib/bpf/bpf_tracing.h */
+/*
+ * BPF_SEQ_PRINTF to wrap bpf_seq_printf to-be-printed values
+ * in a structure.
+ */
+#define BPF_SEQ_PRINTF(seq, fmt, args...)				    \
+	({								    \
+		_Pragma("GCC diagnostic push")				    \
+		_Pragma("GCC diagnostic ignored \"-Wint-conversion\"")	    \
+		static const char ___fmt[] = fmt;			    \
+		unsigned long long ___param[] = { args };		    \
+		_Pragma("GCC diagnostic pop")				    \
+		int ___ret = bpf_seq_printf(seq, ___fmt, sizeof(___fmt),    \
+					    ___param, sizeof(___param));    \
+		___ret;							    \
+	})
+
+
+SEC("iter/task")
+int dump_task(struct bpf_iter__task *ctx)
+{
+	struct seq_file *seq = ctx->meta->seq;
+	__u32 seq_num = ctx->meta->seq_num;
+	struct task_struct *task = ctx->task;
+
+	if (seq_num == 0)
+		BPF_SEQ_PRINTF(seq, "[\n");
+
+	if (task == (void *)0) {
+		BPF_SEQ_PRINTF(seq, "\n]\n");
+		return 0;
+	}
+	if (seq_num) {
+		BPF_SEQ_PRINTF(seq, ",\n");
+	}
+
+        __u64 mntns_id = task->nsproxy->mnt_ns->ns.inum;
+
+
+	BPF_SEQ_PRINTF(seq, "{\"tgid\": %d, \"pid\": %d, \"comm\": \"%s\", \"mntns\": %llu", task->tgid, task->pid, task->comm, mntns_id);
+
+	struct container *container_entry;
+	container_entry = bpf_map_lookup_elem(&containers, &mntns_id);
+	if (container_entry) {
+		BPF_SEQ_PRINTF(seq, ", \"container_id\": \"%s\"", container_entry->ContainerID);
+		BPF_SEQ_PRINTF(seq, ", \"kubernetes_namespace\": \"%s\"", container_entry->KubernetesNamespace);
+		BPF_SEQ_PRINTF(seq, ", \"kubernetes_pod\": \"%s\"", container_entry->KubernetesPod);
+		BPF_SEQ_PRINTF(seq, ", \"kubernetes_container\": \"%s\"", container_entry->KubernetesContainer);
+	}
+
+	BPF_SEQ_PRINTF(seq, "}");
+
+	return 0;
+}
+
+char _license[] SEC("license") = "GPL";
```

Comment:
I'm confused by the licensing. The golang files have an Apache license header. This file doesn't have a header and is licensed GPL? Why is that?

Reply:
When I try to use another license, I get the following from the verifier:
```
117: (85) call bpf_seq_printf#126
cannot call GPL-restricted function from non-GPL compatible program
```

I will add a header.